Improve checking for FHS paths in udev rules

aszlig opened the pull request:
Currently the check against FHS paths in the rule files is only checking against the original paths from in `services.udev.packages`.

However we do fix up some of these paths in the udev rules generator and the warning is against the unfixed rule files and therefore prints a lot of false positives.

This pull request improves the warning and also adds a NixOS option to turn this warning into a failure, so we can use this for VM tests for example and might also be useful for real systems.

Addresses #12722 as well so we can assure that this won't happen again in the future.

Cc: @edolstra

aszlig committed:
nixos/udev: Provide a better warning for FHS paths

We were trying to find FHS references in all of the rules found in
services.udev.packages. Unfortunately we're still fixing up paths in the
same derivation where we are checking those references, so for example
references to /sbin/modprobe were still printed to be needed to fixup
even though they were already fixed at the time.

So now we're printing a more helpful warning message which is also
conditional (before the warning message was printed regardless of
whether there are any rules that need fixup) and is based off the rules
that were already fixed up.

The new warning message not only contains the build-local rule files but
also the original files from other store paths and the FHS path
references that were still found.

With 8ecd3a5e1db40d308d60b5ccfa40eabccb419be9 reverted, we now get this:

/nix/store/...-udev-rules/63-md-raid-arrays.rules (originally from
 /nix/store/...-mdadm-3.3.4/lib/udev/rules.d/63-md-raid-arrays.rules)
 contains references to /usr/bin/readlink and /usr/bin/basename.

Which is now more accurate to what is not yet fixed and where it's
coming from.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/udev: Add an option to fail on FHS paths

So far we were merely printing a warning if there are still references
to (/usr)/s?bin, but we actually want to make sure that we fix those
paths, especially on updates of packages that come with udev rules.

This adds a new option allowImpurePaths, which when set to false will
cause the "udev-rules" derivation to fail.

I've set this to true by default, to not break existing systems too much
and the intention is to set it to false for a few NixOS VM tests.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/tests/installer: Fail on impure udev rules

With 9f2a61c in place, let's actually use this in the installer tests to
make sure we won't shovel FHS paths down the throad of unstable channel
users.

I've tested this by running all of the installer tests for x86_64-linux
and they all succeeded.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

edolstra commented:
Do we need that `allowImpurePaths` option? Maybe we can just fail unconditionally?

aszlig commented:
@edolstra: We don't and I thought about doing this by default as well, but didn't want to cause massive breakages for the systems out there... But maybe we should start to break those systems so we get rid of these paths, what do you think?

aszlig committed:
nixos/udev: Always fail if rules contain FHS paths

Partially reverts the following commits:

  9f2a61c59cc4e4ce278e6582cb4bdca9c2088755
  9c13fe6604358e5255457422acbe8e03734f1e44

As @edolstra pointed out, it would make more sense to do this by default
instead of having that allowImpurePaths option. This of course might
break systems which add extra packages to udev, but on the upside it's
hard to miss one of these paths now because it won't get buried in the
ocean of build output lines.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

edolstra commented:
Yeah, I think this is fine for the master branch.

aszlig commented:
@edolstra: Okay done. Fine if I merge this into master?
